Documentation/guides/MSBuildBestPractices.md

```diff
@@ -308,6 +308,61 @@ Unfortunately, not all targets will have a `$(XDependsOn)` property.
 In some cases, `BeforeTargets` or `AfterTargets` is the only option.
 Consider what happens if the target fails in that case.
 
+## Caching in MSBuild Tasks
+
+There are two scenarios where you may want to do some kind of
+in-memory caching:
+
+1. You have some data you need to share between two MSBuild tasks--or
+   perhaps an instance of the same MSBuild task in another project
+   within the current build.
+2. You have some data you want to persist in-memory between *builds*,
+   while the same instance of the IDE is open.
+
+Scenario 1 can be achieved with `static` variables, although as
+mentioned on [Don't use Static in C#][static_csharp], shared `static`
+state can have its own pitfalls. Avoiding `static` for caching is
+probably a good idea.
+
+Instead we can use an API provided by MSBuild:
+
+```csharp
+// To cache a value
+string key = "foo";
+BuildEngine4.RegisterTaskObject (key, "bar", RegisteredTaskObjectLifetime.Build, allowEarlyCollection: false);
+// To retrieve a value
+string value = (string)BuildEngine4.GetRegisteredTaskObject (key, RegisteredTaskObjectLifetime.Build);
+```
+
+Or if you want to cache across multiple-builds, use
+`RegisteredTaskObjectLifetime.AppDomain` instead.
+
+To *test* and validate your MSBuild task's use of
+`RegisteredTaskObjectLifetime.AppDomain` you have two choices:
+
+* Test builds in the IDE. You can verify with diagnostic logging
+  turned on or using the [Project System Tools][project_system] to
+  view `.binlog` files from builds in the IDE.
+* Set the `%MSBUILDNOINPROCNODE%` environment variable to `1` and
```

Comment:
This doesn't appear to work on Mac at the moment. So we should probably make a note of that.